crates/ra_assists/src/assists/flip_trait_bound.rs

```diff
@@ -0,0 +1,184 @@
+//! `flip_trait_bound` provides an assist that flips the positions of adjacent
+//! type bounds in a type bound list.
+
+use crate::{Assist, AssistCtx, AssistId};
+use hir::db::HirDatabase;
+use ra_syntax::ast::{AstNode, TypeBound, TypeBoundList};
+use ra_syntax::{NodeOrToken, TextRange};
+use std::cmp::{max, min};
+
+/// Flip trait bound assist.
+pub(crate) fn flip_trait_bound(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    // The assist should only be applied when the cursor is in a TypeBoundList
+    let bound_list = ctx.node_at_offset::<TypeBoundList>()?;
+    let bound_list_range = bound_list.syntax().text_range();
+    if !ctx.frange.range.is_subrange(&bound_list_range) {
+        return None;
+    }
+
+    let (bound_at_cursor, other_bound) = type_bound_pair_at_cursor(ctx.frange.range, &bound_list)?;
+
+    let original_cursor_range = ctx.frange.range;
+    ctx.add_action(AssistId("flip_trait_bound"), "flip trait bound", |edit| {
+        let bound_at_cursor_range = bound_at_cursor.syntax().text_range();
+        let other_bound_range = other_bound.syntax().text_range();
+        edit.target(TextRange::from_to(
+            min(bound_at_cursor_range.start(), other_bound_range.start()),
+            max(bound_at_cursor_range.end(), other_bound_range.end()),
+        ));
+        edit.replace(bound_at_cursor_range, other_bound.syntax().text());
+        edit.replace(other_bound_range, bound_at_cursor.syntax().text());
+        edit.set_cursor(original_cursor_range.start());
+    });
+
+    ctx.build()
+}
+
+/// Find the pair of type bounds that should be flipped given `cursor_range`.
+/// - If there are fewer than two TypeBounds in the list, return None.
+/// - If the cursor is between two TypeBounds, these are flipped.
+/// - If the cursor is on a TypeBound, flips with the preceding TypeBound, or the
+/// following one if the cursor is on the last TypeBound in the list.
+fn type_bound_pair_at_cursor(
```

Comment:
I think it would be eaiser to get bounds by starting with a `+` token and using `prev_sibling` and `next_sibling` methods.